### PR TITLE
REFACTOR-#3519: align `mask` interface with its docstring

### LINKED ISSUE #3519: Align backend's masking API with its documentation
Continuous of the discussion we had at the modin discuss:
https://discuss.modin.org/t/does-the-backends-masking-interface-support-slice-usage/227

Solving of this will ease merging of the `iloc/loc` optimizations done in #3513

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

Implementation of the idea discussed at the [modin discuss](https://discuss.modin.org/t/does-the-backends-masking-interface-support-slice-usage/227).

This PR aligns masking implementation via `loc/iloc` with the written docstrings. This PR gets rid of passing slices to the backend as indexers, instead of them, it passes `pandas.RangeIndex` which is list-like (doc-string requires this) and has all the benefits of slices at the same time.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3519 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date 

## Performance changes

With the following ASV benchmarks added:
<details><summary>extra benchmarks</summary>

```python
class TimeIndexingColumns:
    param_names = ["shape"]
    params = [
        get_benchmark_shapes("TimeIndexing")
    ]

    def setup(self, shape):
        self.df = generate_dataframe(ASV_USE_IMPL, "int", *shape, RAND_LOW, RAND_HIGH)
        self.numeric_indexer = [0, 1]
        self.labels_indexer = self.df.columns[self.numeric_indexer].tolist()
    
    def time_iloc(self, shape):
        execute(self.df.iloc[:, self.numeric_indexer])
    
    def time_loc(self, shape):
        execute(self.df.loc[:, self.labels_indexer])
    
    def time___getitem__(self, shape):
        execute(self.df[self.labels_indexer])
```

</details>

Results are the following:

```
      before           after         ratio
     [929c071b]       [711677c9]
     <master>       <range_mask>
         10.4±4ms         12.2±2ms    ~1.17  benchmarks.TimeIndexing.time_loc([1000000, 10], 'scalar')
         20.6±2ms         22.4±2ms     1.09  benchmarks.TimeIndexingColumns.time___getitem__([1000000, 10])
       31.9±0.3ms         34.3±3ms     1.08  benchmarks.TimeIndexing.time_loc([5000, 5000], 'list')
       22.6±0.3ms         24.0±1ms     1.06  benchmarks.TimeIndexing.time_iloc([5000, 5000], 'list')
         17.2±2ms         18.1±1ms     1.05  benchmarks.TimeIndexing.time_iloc([1000000, 10], 'scalar')
          235±3ms          247±3ms     1.05  benchmarks.TimeIndexing.time_iloc([5000, 5000], 'function')
       20.8±0.8ms         21.8±1ms     1.05  benchmarks.TimeIndexing.time_iloc([5000, 5000], 'scalar')
         20.6±2ms         21.0±2ms     1.02  benchmarks.TimeIndexingColumns.time_loc([1000000, 10])
          638±3ms          634±1ms     0.99  benchmarks.TimeIndexing.time_iloc([1000000, 10], 'list')
          782±8ms          778±5ms     0.99  benchmarks.TimeIndexing.time_iloc([1000000, 10], 'function')
       1.11±0.01s       1.10±0.01s     0.99  benchmarks.TimeIndexing.time_loc([1000000, 10], 'list')
          314±6ms          312±8ms     0.99  benchmarks.TimeIndexing.time_loc([1000000, 10], 'bool')
       11.9±0.4ms       11.7±0.6ms     0.99  benchmarks.TimeIndexingColumns.time_loc([5000, 5000])
          392±7ms          385±6ms     0.98  benchmarks.TimeIndexing.time_iloc([1000000, 10], 'bool')
       29.3±0.9ms       28.5±0.5ms     0.97  benchmarks.TimeIndexing.time_loc([5000, 5000], 'scalar')
       12.5±0.5ms       12.2±0.5ms     0.97  benchmarks.TimeIndexingColumns.time___getitem__([5000, 5000])
          459±7ms          444±3ms     0.97  benchmarks.TimeIndexing.time_loc([1000000, 10], 'slice')
          149±6ms          144±9ms     0.97  benchmarks.TimeIndexing.time_loc([5000, 5000], 'bool')
          146±6ms          141±2ms     0.97  benchmarks.TimeIndexing.time_iloc([5000, 5000], 'bool')
         968±20ms         931±20ms     0.96  benchmarks.TimeIndexing.time_loc([1000000, 10], 'function')
          148±7ms          142±7ms     0.96  benchmarks.TimeIndexing.time_loc([5000, 5000], 'slice')
       13.5±0.3ms       12.8±0.4ms     0.95  benchmarks.TimeIndexingColumns.time_iloc([5000, 5000])
         329±10ms          313±5ms     0.95  benchmarks.TimeIndexing.time_loc([5000, 5000], 'function')
          141±8ms          134±8ms     0.95  benchmarks.TimeIndexing.time_iloc([5000, 5000], 'slice')
         23.0±2ms         21.4±1ms     0.93  benchmarks.TimeIndexingColumns.time_iloc([1000000, 10])
-         245±4ms          167±3ms     0.68  benchmarks.TimeIndexing.time_iloc([1000000, 10], 'slice')
```
Overall, performance seems to stay the same.

Extra benchmarks of `apply_func_to_indices_both_axis` with slice indexers, shows that the performance depending on caches hasn't changed:

<details><summary>Ray engine</summary>

|branch|10mln rows, 10 cols (has cache, 112 partitions) | 10mln rows, 10 cols (no cache, 112 partitions) | 1024x1024 (has cache, 1024 partitions) | 1024x1024 (no cache, 1024 partitions) |
|-|-|-|-|-|
|master| 6.23 | 6.21 | 0.10 | 0.08 |
|#3520| 0.63 | 0.78 | 0.09 | 0.06 |

|branch|10mln rows, 10 cols (has cache, 112 partitions) | 10mln rows, 10 cols (has neither frame's nor partition's cache, 112 partitions) | 1024x1024 (has cache, 1024 partitions) | 1024x1024 (has neither frame's nor partition's cache, 1024 partitions) |
|-|-|-|-|-|
|master| 6.07 | 10.53 | 0.10 | 0.43 |
|#3520| 0.60 | 6.22 | 0.10 |  0.32 |

</details>

<details><summary>Dask engine</summary>

|branch|10mln rows, 10 cols (has cache, 112 partitions) | 10mln rows, 10 cols (no cache, 112 partitions) | 1024x1024 (has cache, 1024 partitions) | 1024x1024 (no cache, 1024 partitions) |
|-|-|-|-|-|
|master| 4.71 | 5.04 | 0.23 | 0.24 |
|#3520| 0.85 | 1.07 | 0.24 | 0.24 |

|branch|10mln rows, 10 cols (has cache, 112 partitions) | 10mln rows, 10 cols (has neither frame's nor partition's cache, 112 partitions) | 1024x1024 (has cache, 1024 partitions) | 1024x1024 (has neither frame's nor partition's cache, 1024 partitions) |
|-|-|-|-|-|
|master| 4.64 | 11.21 | 0.24 | 0.84 |
|#3520| 0.85 | 7.34 | 0.25 |  0.82 |

</details>

<details><summary>Reproducer</summary>

```python
import modin.pandas as pd
import numpy as np
from timeit import default_timer as timer

NROWS = 10_000_000
NCOLS = 10

def check_cache(df):
    frame = df._query_compiler._modin_frame
    has_width_cache = 0
    has_length_cache = 0
    for row in frame._partitions:
        for part in row:
            if part._length_cache is not None:
                has_length_cache += 1
            if part._width_cache is not None:
                has_width_cache += 1
    print(f"Total amount of partitions: {frame._partitions.shape[0] * frame._partitions.shape[1]}")
    print(f"\tNumber of partitions having width cache: {has_width_cache}")
    print(f"\tNumber of partitions having length cache: {has_length_cache}")

def check_frame_cache(df):
    frame = df._query_compiler._modin_frame
    print(f"Frame's column widths cache: {frame._column_widths_cache}")
    print(f"Frame's row lengths cache: {frame._row_lengths_cache}")

df = pd.DataFrame({f"col{i}": np.arange(NROWS) for i in range(NCOLS)})

value = df.values

check_cache(df)
check_frame_cache(df)

t1 = timer()
df.iloc[:, :] = value
repr(df)
print(f"has cache: {timer() - t1}")

# Axis relabeling resets partition's cache
df.index = df.index
df._query_compiler._modin_frame._column_widths_cache = None
df._query_compiler._modin_frame._row_lengths_cache = None
check_cache(df)
check_frame_cache(df)

t1 = timer()
df.iloc[:, :] = value
repr(df)
print(f"no cache: {timer() - t1}")

# Possible output:
# Total amount of partitions: 112
#         Number of partitions having width cache: 112
#         Number of partitions having length cache: 112
# Frame's column widths cache: [10]
# Frame's row lengths cache: [89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89286, 89254]
# has cache: 0.6110919890925288
# Total amount of partitions: 112
#         Number of partitions having width cache: 0
#         Number of partitions having length cache: 0
# Frame's column widths cache: None
# Frame's row lengths cache: None
# no cache: 6.532485898584127


```

</details>

You can find more info about why it is required to test in [this thread](https://github.com/modin-project/modin/pull/3520#discussion_r736543884).

### CODE REVIEW

File: modin/backends/base/query_compiler.py [hunk 1]
```diff
@@ -2548,8 +2548,8 @@ def view(self, index=None, columns=None):
         BaseQueryCompiler
             New masked QueryCompiler.
         """
-        index = [] if index is None else index
-        columns = [] if columns is None else columns
+        index = slice(None) if index is None else index
+        columns = slice(None) if columns is None else columns
```
Comment (dchigarev) [inline modin/backends/base/query_compiler.py:2552]: bug in the `BaseQueryCompiler`. We have to grab the whole axis if the indexer is None, previously it was taking no elements instead.

File: modin/engines/base/frame/data.py [hunk 2]
```diff
@@ -467,19 +472,24 @@ def mask(
...
+                    # length, but sometimes it is different, thus the extra calculation.
+                    range(*idx.indices(self._row_lengths[p]))
+                    if isinstance(idx, slice)
+                    else idx
+                )
+                for p, idx in row_partitions_list.items()
+            ]
+            new_index = self.index[
+                # Pandas Index is more likely to preserve its metadata if the indexer is slice
+                slice(row_numeric_idx.start, row_numeric_idx.stop, row_numeric_idx.step)
```
Comment (dchigarev) [inline modin/engines/base/frame/data.py:488]: for example:
```python
>>> index = pandas.date_range("2017/1/4", periods=10)
>>> repr(index[slice(2, 5)].freq)
'<Day>'
>>> repr(index[range(2, 5)].freq)
'None'
```

File: modin/pandas/indexing.py [hunk 3]
```diff
@@ -799,32 +812,30 @@ def _compute_lookup(self, row_loc, col_loc):
...
-        ):
-            col_lookup = (
-                pandas.RangeIndex(len(self.qc.columns)).to_series().iloc[col_loc].index
-            )
-        else:
-            col_lookup = col_loc
-        return row_lookup, col_lookup
+        lookups = []
+        for axis, axis_loc in enumerate((row_loc, col_loc)):
+            if isinstance(axis_loc, slice) and axis_loc.step is None:
```
Comment (dchigarev) [inline modin/pandas/indexing.py:822]: the same condition as before but negated in order to make it simpler

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 4]
```diff
@@ -1148,19 +1152,31 @@ def apply_func_to_indices_both_axis(
         """
         partition_copy = partitions.copy()
         row_position_counter = 0
+
+        def compute_part_size(internal_idx, remote_part, axis):
+            """Compute indexer length along the specified axis for the passed partition."""
+            if isinstance(internal_idx, slice):
+                internal_idx = range(
+                    *internal_idx.indices(
+                        remote_part.width() if axis == 1 else remote_part.length()
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: added benchmarks for Dask engine to the head of the PR

**Review by vnlitvinov — APPROVED** [review]
LGTM

**Review by YarShev — COMMENTED** [review]
Overall, LGTM! Left a few comments.

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 5]
```diff
@@ -446,15 +447,25 @@ def mask(
...
-            row_numeric_idx = None
-        if isinstance(col_numeric_idx, slice) and (
-            col_numeric_idx == slice(None) or col_numeric_idx == slice(0, None)
-        ):
-            col_numeric_idx = None
+        indexers = []
+        for axis, indexer in enumerate((row_numeric_idx, col_numeric_idx)):
+            if is_range_like(indexer):
+                if indexer.step == 1 and len(indexer) == len(self.axes[axis]):
+                    indexer = None
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:455]: added the following in-code comment:
"By semantic of this function `None` indexer is a full axis access"

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 6]
```diff
@@ -927,13 +943,21 @@ def _get_dict_of_block_index(self, axis, indices):
             A mapping from partition index to list of internal indices which correspond to `indices` in each
             partition.
         """
+        # TODO: Support handling of slices with specified 'step'. For now, converting them into a range
+        if isinstance(indices, slice) and (
+            indices.step is not None and indices.step != 1
+        ):
+            list_of_indices = range(*indices.indices(len(self.axes[axis])))
+            return self._get_dict_of_block_index(axis=axis, indices=list_of_indices)
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: we don't need it. removed

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 7]
```diff
@@ -495,36 +511,34 @@ def mask(
...
-                new_columns = self.columns[sorted(col_numeric_idx)]
-                if self._dtypes is not None:
-                    new_dtypes = self.dtypes.iloc[sorted(col_numeric_idx)]
-                else:
-                    new_dtypes = None
+                monotonic_col_idx = sorted(col_numeric_idx)
+            new_columns = self.columns[monotonic_col_idx]
+            assert sum(new_col_widths) == len(
+                new_columns
+            ), f"{sum(new_col_widths)} != {len(new_columns)}.\n{col_numeric_idx}\n{self._column_widths}\n{col_partitions_list}"
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: This error may occur if something wrong happened with the partitioning (for example it became unaligned), so it's useful to log partitions column widths. I kept the error message untouched but replaced a pure assert with the `ErrorMessage`.

BTW, I haven't added this assert in this PR, it is presented in the current master:
https://github.com/modin-project/modin/blob/7ab2d909d1ad42dd0b9d550e2b4e78fd03abf863/modin/core/dataframe/pandas/dataframe/dataframe.py#L508-L516

**Comment by dchigarev** [comment]
@YarShev applied suggestions and CI is green

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 8]
```diff
@@ -947,8 +949,7 @@ def _get_dict_of_block_index(self, axis, indices):
         if isinstance(indices, slice) and (
             indices.step is not None and indices.step != 1
         ):
-            list_of_indices = range(*indices.indices(len(self.axes[axis])))
-            return self._get_dict_of_block_index(axis=axis, indices=list_of_indices)
```
Comment (vnlitvinov) [inline modin/core/dataframe/pandas/dataframe/dataframe.py]: why `return` was removed?

Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: I just removed a recursive call of `_get_dict_of_block_index`, it appeared, that there is no sense in computing a new `indexer` parameter and calling the function again if we can just process a non-one-step indexer here and continue the execution

**Comment by dchigarev** [comment]
@YarShev @vnlitvinov all the comments have been addressed and CI has passed

File: modin/core/dataframe/pandas/dataframe/dataframe.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: ok, I see

**Review by YarShev — APPROVED** [review]
LGTM

**Comment by dchigarev** [comment]
@vnlitvinov a reminder to review

**Review by vnlitvinov — APPROVED** [review]
LGTM, thanks!

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 9]
```diff
@@ -446,15 +447,26 @@ def mask(
         If both `row_indices` and `row_numeric_idx` are set, `row_indices` will be used.
         The same rule applied to `col_indices` and `col_numeric_idx`.
         """
-        # Check on all possible ranges
```
Comment (jbrockmendel) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:449]: @dchigarev the analogous method in cuDFOnRayDataframe still uses the old code here.  any reason not to update there too?

Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:449]: It seems that I've just missed `cuDFOnRayDataframe` when I did this PR, so basically, there's no reason not to propagate the same logic to the [`cuDFOnRayDataframe.take_2d_labels_or_positional`](https://github.com/modin-project/modin/blob/dff84b9cdf373096f88356fe17202599c192c2fc/modin/core/execution/ray/implementations/cudf_on_ray/dataframe/dataframe.py#L120)

**Pull request merged**